[Minor Release] - Transfer Family Web App Modules and Examples

oliviamelaniputrii opened the pull request:
## Add AWS Transfer Family Web App Module and Example

### Overview
This PR introduces a new Transfer Web App module and comprehensive example that enables browser-based file transfer with AWS IAM Identity Center authentication and S3 Access Grants for fine-grained permissions.

### What's Added

#### New Module: modules/transfer-web-app
- **Browser-based interface** for secure S3 file access
- **Identity Center integration** with SSO authentication support
- **S3 Access Grants** for user and group-level permissions
- **Custom branding** support (logo, favicon, title)
- **Built-in validation** for proper configuration
- **Flexible S3 Access Grants configuration** (new or existing instances/locations)

#### New Example: examples/sftp-web-app
- **Complete end-to-end deployment** with Identity Center, S3, CloudTrail, and web app
- **Flexible user/group management** - create new or import existing Identity Center users/groups
- **CloudTrail audit logging** with KMS encryption and SNS notifications
- **CORS configuration** restricted to web app endpoint
- **Custom branding** with AnyCompany logo, favicon, and title as defaults
- **Built-in validation** with multi-layer permission checks

### Key Features
- **Flexible deployment**: Supports both test user creation and existing Identity Center integration
- **Validation system**: Multi-layer validation catches configuration errors at plan time
- **Documentation**: Complete documentation with examples, configuration guides, and best practices

oliviamelaniputrii commented on .header.md:
```diff
@@ -177,6 +178,19 @@ This project utilizes multiple modules to create a complete AWS Transfer Family
+  - S3 Access Grants for fine-grained permissions
+  - Custom branding with logo, favicon, and title
+  - CORS configuration support
+  - Built-in validation for proper configuration
```
for the example and module which include the web app - as we have discussed during our meeting

oliviamelaniputrii commented on .header.md:
```diff
@@ -393,6 +407,22 @@ Key features:
 - Configurable destination buckets for clean, infected, and error files
 - Optional file cleanup from ingestion bucket
 
+## Example for SFTP Web App
```
@prabirsekhri @pururajsinh-amazon any suggestion on the naming to replace sftp-web-app and the corresponding phrases for the example? I am not inclined to use transfer-web-app since that is for the module

pururajsinh-amazon commented on .header.md:
```diff
@@ -393,6 +407,22 @@ Key features:
+Key features:
+
+- Browser-based interface for secure file access
+- AWS IAM Identity Center authentication with SSO support
```
Not sure if we have to mention SSO since the new name is identity center, any chance we can make the verbiage identical to the module for Identity center?

oliviamelaniputrii commented on examples/sftp-web-app/.header.md:
```diff
@@ -0,0 +1,154 @@
+
+- `aws_region` (default: `"us-east-1"`) - AWS region for deployment
+- `identity_center_instance_arn` (default: `null`) - ARN of Identity Center instance, uses first available if not specified
+- `create_identity_center_instance` (default: `false`) - Whether to create a new Identity Center instance
```
it'll be the account level i believe, anything default created by this https://registry.terraform.io/providers/hashicorp/awscc/latest/docs/data-sources/sso_instance

pururajsinh-amazon replied:
okay, we can clarify this in the documentation

oliviamelaniputrii replied:
since this is defaulted to be false and mostly are for testing, should we even emphasize and call it out?

oliviamelaniputrii commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+#####################################################################################
+# Create IAM Identity Center instance (optional)
+#####################################################################################
+resource "awscc_sso_instance" "identity_center" {
```
I dont think we can do it through terraform https://registry.terraform.io/providers/hashicorp/awscc/latest/docs/resources/sso_instance

pururajsinh-amazon replied:
Also should this be the first thing in the main.tf - see comment above.

oliviamelaniputrii replied:
ordering does not matter on the creation, just want to standardize the code organization to put all the checks, defaults and locals at the beginning, wdyt?

pururajsinh-amazon commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+
+module "s3_bucket_cloudtrail_logs" {
+  source                   = "git::https://github.com/terraform-aws-modules/terraform-aws-s3-bucket.git?ref=v5.0.0"
+  bucket                   = lower("${random_pet.name.id}-cloudtrail-logs-${random_id.suffix.hex}")
```
should this contain `web-app` in the name also?

oliviamelaniputrii commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+          permission = grant.permission
+        }
+      ]
+    }
```
it's in the module

pururajsinh-amazon replied:
thanks for clarifying. Newbie terraform question - How does it work though? The main.tf for the example is not having references to application assignments?🤔

oliviamelaniputrii replied:
things will be handled by the module, that's why we are passing the username and group name from the example to the module. then module will take it and do the assignment

pururajsinh-amazon commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+        Sid    = "Enable IAM User Permissions"
+        Effect = "Allow"
+        Principal = {
+          AWS = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
```
`aws` will be different based on the partition cn, us-gov, eusc etc. We should handle that based on the region info

pururajsinh-amazon commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+        Resource = aws_kms_key.cloudtrail.arn
+        Condition = {
+          ArnEquals = {
+            "kms:EncryptionContext:aws:logs:arn" = "arn:aws:logs:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:log-group:/aws/cloudtrail/${random_pet.name.id}"
```
same comment about the partition in the arn

pururajsinh-amazon replied:
responded

oliviamelaniputrii commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+      }
+    ]
+  })
+}
```
i think it's just github, no error on my IDE

ulrokx commented on .header.md:
```diff
@@ -393,6 +407,22 @@ Key features:
+- CloudTrail audit logging with KMS encryption and SNS notifications
+- Custom branding with logo, favicon, and title customization
+- Flexible user/group management (create new or import existing)
+- CORS configuration for secure web app access
```
While CORS is a security feature, we might give the wrong impression here that setting this is optional which can lead to a hard-to-debug issue with the web app. Maybe this could be rephrased as `Required S3 CORS configuration for secure web app access`

pururajsinh-amazon commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+#####################################################################################
+# Create Amazon S3 destination bucket for file storage
+#####################################################################################
+module "s3_bucket" {
```
not sure I follow, but can discuss this offline

pururajsinh-amazon commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+        Principal = {
+          AWS = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
+        }
+        Action   = "kms:*"
```
Is a root user expected to do kms:* by assuming this role?

pururajsinh-amazon commented on examples/sftp-web-app/main.tf:
```diff
@@ -0,0 +1,485 @@
+  name                          = "${random_pet.name.id}-audit-trail"
+  s3_bucket_name                = module.s3_bucket_cloudtrail_logs.s3_bucket_id
+  include_global_service_events = true
+  is_multi_region_trail         = true
```
Sorry I need more context here, is it expected to ship the trails cross region? What is checkov complaining about?

oliviamelaniputrii replied:
yeah checkov expected it to be set true - I tried different approach which might allow us to omit the   include_global_service_events = true and  is_multi_region_trail         = true

will test them out again

oliviamelaniputrii commented on examples/sftp-web-app/anycompany-logo-small.png:
note: checked with Suh and Prabir

pururajsinh-amazon commented on modules/transfer-custom-idp-solution/.header.md:
```diff
@@ -0,0 +1,153 @@
+# AWS Transfer Family Custom Identity Provider Solution Module
```
This is not supposed to be reviewed, right?

pururajsinh-amazon commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,415 @@
+}
+
+locals {
+  identity_store_id            = tolist(data.aws_ssoadmin_instances.identity_center.identity_store_ids)[0]
```
same comment about the lookups as called out earlier

oliviamelaniputrii commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,415 @@
+}
+
+# IAM policy document for S3 Access Grants
+data "aws_iam_policy_document" "transfer_web_app" {
```
Yeah they can modify the code themselves - wdyt?

pururajsinh-amazon replied:
yea

pururajsinh-amazon commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,415 @@
+
+# Assign users to Transfer Family Web App via Identity Center Application
+resource "aws_ssoadmin_application_assignment" "users" {
+  for_each = { for user in var.identity_center_users : user.username => user }
```
I think we will probably think about the TPS for application assignment and create users / groups since they have a defined TPS. 
I am not sure if terraform explicitly handles this or if we need to have customer implementation to take care of it. I think its the later but probably Olivia can confirm this.
https://docs.aws.amazon.com/singlesignon/latest/userguide/limits.html

pururajsinh-amazon replied:
If our end to end tests are creating module with 50-100 users and groups, we will get some good data points.

pururajsinh-amazon commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,415 @@
+    sid    = "BucketLevelReadPermissions"
+    effect = "Allow"
+    actions = [
+      "s3:ListBucket"
```
This would be s3:ListBuckets

pururajsinh-amazon commented on modules/transfer-web-app/outputs.tf:
```diff
@@ -0,0 +1,41 @@
+# Outputs for Transfer Web App Module
+
```
Should we call out the AccessGrants roles in the output?

oliviamelaniputrii commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,415 @@
+}
+
+# IAM role for Transfer web app
+resource "aws_iam_role" "transfer_web_app" {
```
Do we want to make the existing role option as well for the web app role? Made this an option for s3 ag but thinking of being opiniated on web app

pururajsinh-amazon replied:
sorry what is the count doing?

oliviamelaniputrii commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,415 @@
+
+# IAM role for Transfer web app
+resource "aws_iam_role" "transfer_web_app" {
+  name               = var.iam_role_name
```
changes applied - please check if this is correct @pururajsinh-amazon

oliviamelaniputrii commented:
/do-e2e-tests

oliviamelaniputrii commented:
/do-e2e-tests

oliviamelaniputrii commented:
/do-e2e-tests

prabirsekhri commented on modules/transfer-web-app/main.tf:
```diff
@@ -0,0 +1,420 @@
+##################################################################
+# AWS Transfer Family Web App Module
+# This module web application resources for AWS Transfer Family
```
Is there a typo? Did you mean?  This module creates web apps resources?

oliviamelaniputrii committed:
feat(web-app): Add Transfer Family Web App module

Add new transfer-web-app module with the following capabilities:
- Browser-based interface for secure S3 file access
- AWS IAM Identity Center integration for authentication
- S3 Access Grants for fine-grained user and group permissions
- Custom branding support (logo, favicon, title)
- Flexible configuration for new or existing S3 Access Grants instances
- Built-in validation for proper configuration

oliviamelaniputrii committed:
feat(example): Add sample-web-app example with Identity Center integration

Add comprehensive example demonstrating Transfer Family Web App deployment:
- Complete end-to-end setup with IAM Identity Center users/groups
- CloudTrail integration with KMS encryption and SNS notifications
- S3 bucket with CORS configuration for web app access
- Flexible user/group management (create new or import existing)
- Custom branding with logo and favicon support
- Built-in validation for configuration errors

oliviamelaniputrii committed:
docs: Add documentation for Web App module and update root README

- Add .header.md and README.md for transfer-web-app module
- Add .header.md and README.md for sample-web-app example
- Update root .header.md and README.md with Web App module info
- Add .header.md for transfer-custom-idp-solution module
- Update transfer-custom-idp-solution README with terraform-docs format
- Add AWS partition note for commercial regions testing

oliviamelaniputrii committed:
test: Add Web App module tests and fix connector test configuration

- Add plan and apply tests for sample-web-app example
- Disable DynamoDB tracking in connector retrieve tests
- Configure test variables for Identity Center instance creation

prabirsekhri reviewed (approved):
LGTM
